### PR TITLE
1151 http refactor

### LINKED ISSUE #1151: Refactor http handler
### Description
Currently there is a lot of Pilosa logic embedded in handler.go. If we want to be able to do things like embedding pilosa #1057 or having alternate interfaces such as GRPC #642, we need to move common logic out of handler.go. 

For example, let's take `handlePostIndex`:
1. gets index name out of the http path
2. decodes the request body
3. calls holder.CreateIndex
4. calls Broadcaster.SendSync
5. encodes a response
6. increments stats

If we just make the handler responsible for HTTP, we might have a function which looks like `func CreateIndex(index string, body io.Reader) error` This function would decode the request body (content type info may need to be passed along, but the encoding (pbuf or json) should be distinct from the transport (http in this case)). Then it would create the index, call broadcaster.sendsync, and update the stats. The postIndexResponse is empty, but one could imagine this function returning that along with the error. The other missing piece is making sure that the correct HTTP error code is returned in the case of an error. This will depend on us implementing some notion of error codes in Pilosa and then having a mapping of Pilosa errors onto HTTP errors (or any other transport).

I think this work can be done a few endpoints at a time to avoid giant PRs full of merge conflicts.

Once it's done, we can move all http handling code into a subpackage, in the spirit of #1093 (and as suggested by Ben Johnson's package management strategy).

### Success criteria (What criteria will consider this ticket closeable?)
handler.go is responsible for http handling and little else. It no longer imports encoding/json, protobuf, pilosa/pql, pilosa/internal, etc.

### PR DESCRIPTION
**Pull request by jaffee**
## Overview

Rebased yuce's 1151 work onto master after cluster-resize was merged. Lot's of conflicts and work needed to be done to get tests passing. Concerned that untested functionality may have been missed. 

Examples:
- Recalculate caches gets broadcast to all nodes - this got stomped by the rebase, had to re-add.
- I left the InputJSONDataParser handler commented out, and no tests complain.
- maxslices endpoint changed - had to be redone.
- status/schema endpoints

Fixes #1151 

Update: I'm satisfied with this now, I reviewed the API against all changed to handler.go since v0.8.8 and I don't think we missed anything - let's get it merged ASAP to avoid more merge conflicts. I've identified some follow-on work around errors, and I think /version and /debug/vars need to be pulled into API as well, but I'm of the opinion that we can do that separately.

## Pull request checklist

- [x] I have read the [contributing guide](https://github.com/pilosa/pilosa/blob/master/CONTRIBUTING.md).
- [x] I have agreed to the [Contributor License Agreement](https://cla-assistant.io/pilosa/pilosa).
- [ ] I have updated the [documentation](https://github.com/pilosa/pilosa/tree/master/docs).
- [x] I have resolved any merge conflicts.
- [ ] I have included tests that cover my changes.
- [x] All new and existing tests pass.

## Code review checklist
This is the checklist that the reviewer will follow while reviewing your pull request. You do not need to do anything with this checklist, but be aware of what the reviewer will be looking for.

- [ ] Ensure that any changes to external docs have been included in this pull request.
- [ ] If the changes require that minor/major versions need to be updated, tag the PR appropriately.
- [ ] Ensure the new code is [properly commented](https://github.com/golang/go/wiki/CodeReviewComments#doc-comments) and follows [Idiomatic Go](https://dmitri.shuralyov.com/idiomatic-go).
- [ ] Check that tests have been written and that they cover the new functionality.
- [ ] Run tests and ensure they pass.
- [ ] Build and run the code, performing any applicable integration testing.

### CODE REVIEW

**Comment by yuce** [comment]
I get `rowID required` errors when I go client tests against this branch. Should the `master` merged into this branch?

**Comment by yuce** [comment]
Merging `master` into this branch fixes `rowID required` errors. There wasn't any merge conflicts, all Pilosa server and Go client tests pass.

**Comment by jaffee** [comment]
thanks yuce - I merged master in and pushed that up

File: test/handler.go [hunk 1]
```diff
@@ -40,10 +40,12 @@ func NewHandler() *Handler {
 	h := &Handler{
 		Handler: pilosa.NewHandler(),
 	}
-	h.Handler.Executor = &h.Executor
+	h.API = pilosa.NewAPI()
+	h.Handler.API = pilosa.NewAPI()
```
Comment (travisturner) [inline test/handler.go]: should this be `h.Handler.API = h.API` instead? perhaps it doesn't matter, but I'm wondering if having those separate might cause confusion in the future.

File: handler.go [hunk 2]
```diff
@@ -2133,83 +1734,82 @@ type clusterResizeAbortResponse struct {
...
-				return nil, fmt.Errorf("error handling action: %s, err: %s", action.ValueDestination, err)
-			}
-			if bit != nil {
-				setBits[frame] = append(setBits[frame], bit)
-			}
-		}
-	}
-	return setBits, nil
-}
+// // InputJSONDataParser validates input json file and executes SetBit.
```
Comment (travisturner) [inline handler.go]: any reason to keep this commented code?

Comment (jaffee) [reply handler.go]: Ah yes -  are we OK with aggressively deprecating input definition? This is basically just me being lazy, although the fact that all our tests pass without this code is probably indicative of something...

File: test/handler.go (hunk 1 above)
Comment (jaffee) [reply test/handler.go]: I think you're right

File: handler.go (hunk 2 above)
Comment (travisturner) [reply handler.go]: it looks like that method became `inputJSONDataParser` in `api.go`

Comment (jaffee) [reply handler.go]: ohh, I see what happened now. For some reason I thought that the handler just hadn't been ported, but it was a helper function sneaking back in. I removed it.

**Pull request merged**